lib/srv/app/cloud_test.go

```diff
@@ -0,0 +1,280 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package app
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	awssession "github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+)
+
+func TestIsSessionUsingTemporaryCredentials(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name        string
+		credentials *credentials.Credentials
+		expectBool  bool
+		expectError func(error) bool
+	}{
+		{
+			name: "ec2 role",
+			credentials: credentials.NewCredentials(&mockCredentialsProvider{
+				retrieveValue: credentials.Value{
+					AccessKeyID:     "id",
+					SecretAccessKey: "secret",
+					ProviderName:    ec2rolecreds.ProviderName,
+				},
+			}),
+			expectBool: false,
+		},
+		{
+			name: "web identity",
+			credentials: credentials.NewCredentials(&mockCredentialsProvider{
+				retrieveValue: credentials.Value{
+					AccessKeyID:     "id",
+					SecretAccessKey: "secret",
+					SessionToken:    "token",
+					ProviderName:    stscreds.WebIdentityProviderName,
+				},
+			}),
+			expectBool: true,
+		},
+		{
+			name: "session token exists",
+			credentials: credentials.NewCredentials(&mockCredentialsProvider{
+				retrieveValue: credentials.Value{
+					AccessKeyID:     "id",
+					SecretAccessKey: "secret",
+					SessionToken:    "token",
+					ProviderName:    "SharedConfigCredentials",
+				},
+			}),
+			expectBool: true,
+		},
+		{
+			name:        "bad config",
+			credentials: nil,
+			expectError: trace.IsNotFound,
+		},
+		{
+			name: "failed to get credentials",
+			credentials: credentials.NewCredentials(&mockCredentialsProvider{
+				retrieveError: trace.AccessDenied(""),
+			}),
+			expectError: trace.IsAccessDenied,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			c, err := NewCloud(CloudConfig{
+				Session: &awssession.Session{
+					Config: &aws.Config{
+						Credentials: test.credentials,
+					},
+				},
+			})
+			require.NoError(t, err)
+
+			isTemporary, err := c.(*cloud).isSessionUsingTemporaryCredentials()
```

Comment:
Nit: I know this is a test but I'd probably do a safe type assertion check anyway.